text/0126-Configurable-Metric-Aggregations.md

```diff
@@ -0,0 +1,95 @@
+  - Selection options include: the instrument type (Counter, ValueRecorder, etc), and a regex for instrument name.
+  - If more than one option is provided, they are considered additive.
+  - Example: select all ValueRecorders whose name ends with ".duration".
+* Configuration - configures how the batching and aggregation should be done.
+  - 3 things can be specified: The aggregation (Sum, MinMaxSumCount, Histogram, etc), the "temporality" of the batching,
+    and a set of pre-defined labels to consider as the subset to be used for aggregations.
+    - Note: "temporality" can be one of "DELTA" and "CUMULATIVE" and specifies whether the values of the aggregation
+      are reset after a collection is done or not, respectively.
+  - If not all are specified, then the others should be considered to be requesting the default.
+  - Examples:
+    - Use a MinMaxSumCount aggregation, and provide delta-style batching.
+    - Use a Histogram aggregation, and only use two labels "route" and "error" for aggregations.
+    - Use a quantile aggregation, and drop all labels when aggregating.
+
+In this proposal, there is only one configuration associated with each selector.
+
+As a concrete example, in Java, this might look something like this:
+
+```java
+// get a handle to the MeterSdkProvider
+ MeterSdkProvider meterProvider = OpenTelemetrySdk.getMeterProvider();
+
+ // create a selector to select which instruments to customize:
+ InstrumentSelector instrumentSelector = InstrumentSelector.newBuilder()
+  .instrumentType(InstrumentType.COUNTER)
+  .build();
+
+ // create a specification of how you want the metrics aggregated:
+ Specification specification =
+      specification.create(Aggregations.minMaxSumCount(), Temporality.DELTA);
+
+ //register the configuration with the MeterSdkProvider
+ meterProvider.registerAggregation(instrumentSelector, specification);
+```
+
+## Internal details
+
+This OTEP does not specify how this should be implemented in a particular language, only the functionality that is desired.
+
+A prototype with a partial implementation of this proposal in Java is available in PR form [here](https://github.com/open-telemetry/opentelemetry-java/pull/1412)
+
+## Trade-offs and mitigations
+
+This does not intend to deliver a full "Views" API, although it may eventually form the basis for one. The goal here is
+simply to allow configuration of the batching and aggregation by operators and exporter authors.
+
+This does not intend to specify the exact interface for providing these configurations, nor does it
+consider a non-programmatic configuration option.
+
+## Prior art and alternatives
+
+* Prior Art is probably mostly in the [OpenCensus Views](https://opencensus.io/stats/view/) system.
+* Another [OTEP](https://github.com/open-telemetry/oteps/pull/8) attempted to address building a Views API.
+
+## Open questions
+
+There has been some question about whether a custom aggregation should be allowable for a Counter type
```

Comment:
I'll put a concern of mine with custom aggregators on Counters here with an example, it is Erlang but my have parallels to other languages people will see and provide feedback as well:

An instrument has a kind and a number kind, for this case we'll have an integer counter. If we know that the recordings to this instrument will always be treated as a sum then a potential optimization is available when using `bind_instrument`. 

Erlang has shared mutable counters (https://erlang.org/doc/man/counters.html) so a `bind` could return a reference to the shared mutable counter for instrument `I`. Each `bind` in any process would return the same reference for `I`.

If `I` is a counter but could have any and multiple aggregators that reference it then a `bind` could not return a shared mutable counter but must record the value the same as with `ValueRecorder` instruments.

I suppose that technically at the time of `bind` the aggregators configured for `I` could be checked and if the only existing one is the `Sum` aggregator it can return the shared mutable counter... But it feels like we are wasting valuable information that makes for a more efficient handling of the instruments. In OpenCensus there was only the equivalent of `ValueRecorder` so this wasn't a consideration, with OpenTelemetry I saw opportunity for optimizations.

Reply:
If we truly want to make this restriction, I'm not opposed to having a specific option create an error if your SDK can't support it. For example, in java right now, not all aggregations are supported for all instrument types, so this kind of error seems like a possible escape clause, per SDK.

Reply:
What wasted information is there? The reason I mention the Counter and Histogram configuration is that it's a common scenario where you both want to monitor a size distribution and treat the sum as a metric. Deciding whether to use the faster behavior (default case) or the slower behavior (e.g., histogram) is supposed to be an option, but the default case should be the efficient one.

Reply:
@jmacd the fact that a counter is a counter is only used for setting the default aggregation. It feels more like when creating an instrument the user can just set its default aggregation and no need for multiple instrument kinds. The fact that we have instrument kinds I think gives the impression that the act of recording is specialized for each one.

I don't like the idea of the Erlang/Elixir SDK acting differently from other languages (at least not where it is absolutely required) so if this is the direction the SDK keeps I'll have it act the same as the Erlang OpenCensus implementation where measure recording is basically the same as this.

Reply:
The Counter is not only used to set the default aggregation, it's used to imply ADDING and MONOTONIC which tell you how you can use the information regardless of which aggregation is applied. So if you choose to use a Histogram aggregation with a Counter instrument, you still know that the sum indicates a meaningful rate (because of ADDING and MONOTONIC).

Reply:
I think this question is outside of the scope of this particular OTEP.  The concern about Counters is in the open questions. :)